### PR TITLE
TEST: write ops should execute under shard permit

### PR DESCRIPTION
**Pull request by dnhatn**
Currently `ESIndexLevelReplicationTestCase` executes write operations
without acquiring  index shard permit. This may prevent the primary term
on replica from being updated or cause a race between resync and
indexing on primary. This commit ensures that write operations in the shard-level 
tests are always executed under shard permit like the production code.

### CODE REVIEW

**Review by bleskes — APPROVED** [review]
This is a good one. I do wonder why no test failed as this is an important part of our semantics. I presume this is because we don't have any test that does things concurrently in just the right way. Do you mind double checking and adding such a test if missing?

**Comment by dnhatn** [comment]
@bleskes 
> Do you mind double checking and adding such a test if missing?

Sure, I will do.

**Review by jasontedor — COMMENTED** [review]
I left a few comments.

File: server/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 1]
```diff
@@ -602,15 +602,25 @@ protected void performOnReplica(BulkShardRequest request, IndexShard replica) th
...
+                new TransportShardBulkActionTests.NoopMappingUpdatePerformer());
+        }
         TransportWriteActionTestHelper.performPostWriteActions(primary, request, result.location, logger);
         return result;
     }
 
-    private void executeShardBulkOnReplica(IndexShard replica, BulkShardRequest request) throws Exception {
-        final Translog.Location location = TransportShardBulkAction.performOnReplica(request, replica);
+    private void executeShardBulkOnReplica(ReplicationGroup group, IndexShard replica, BulkShardRequest request) throws Exception {
+        final PlainActionFuture<Releasable> permitAcquirer = new PlainActionFuture<>();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: ++. `permitAcquiredFuture` is better.

**Comment by dnhatn** [comment]
@jasontedor I've addressed your comments. Can you take a look? Thank you.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by dnhatn** [comment]
Thanks @bleskes and @jasontedor for reviewing.

**Comment by bleskes** [comment]
> > Do you mind double checking and adding such a test if missing?
> Sure, I will do.

Yw @dnhatn . I presume ^^ is a follow up?

**Comment by dnhatn** [comment]
> Yw @dnhatn . I presume ^^ is a follow up?

@bleskes Yes!

**Pull request merged**